contracts/RegistryV1.sol

```diff
@@ -0,0 +1,300 @@
+  event RewardsAgreementApproved(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+  event ReferralRegistered(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer,
+    address indexed user
+  );
+  event ReferralFailed(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer,
+    address indexed user,
+    string reason
+  );
+  event RequiresApprovalForRewardsAgreements(
+    address indexed entity,
+    bool requiresApproval
+  );
+
+  // Errors
+  error InvalidEntityAddress(address entity);
+  error EntityAlreadyExists(address entity);
+  error EntityDoesNotExist(address entity);
+  error NotEntityOwner(address entity, address owner);
+  error AgreementAlreadyExists(address provider, address consumer);
+  error ProviderRequiresApproval(address provider);
+  error InvalidSignature();
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address owner, uint48 transferDelay) public initializer {
+    __AccessControlDefaultAdminRules_init(transferDelay, owner);
+    __UUPSUpgradeable_init();
+  }
+
+  function _authorizeUpgrade(
+    address
+  ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {} // solhint-disable-line no-empty-blocks
+
+  /**
+   * @notice Register a new rewards entity
+   * @param entity The entity address to register
+   */
+  function registerRewardsEntity(address entity) external {
+    if (entity == address(0)) {
+      revert InvalidEntityAddress(entity);
+    }
+
+    if (_entityToId[entity] != 0) {
+      revert EntityAlreadyExists(entity);
+    }
+
+    uint256 entityId = ++_nextEntityId;
+    _idToEntity[entityId] = entity;
+    _entityToId[entity] = entityId;
+
+    emit RewardsEntityRegistered(entity);
+  }
+
+  /**
+   * @notice Transfer ownership of a rewards entity
+   * @param entity The entity address
+   * @param newOwner The new owner address
+   */
+  function transferRewardsEntityOwner(
+    address entity,
+    address newOwner
+  ) external {
+    uint256 entityId = _entityToId[entity];
+    if (entityId == 0) {
+      revert EntityDoesNotExist(entity);
+    }
+
+    if (entity != msg.sender) {
+      revert NotEntityOwner(entity, msg.sender);
+    }
+
+    // Update mappings
+    _idToEntity[entityId] = newOwner;
+    _entityToId[newOwner] = entityId;
+    delete _entityToId[entity];
+
+    emit RewardsEntityOwnerTransferred(entity, newOwner);
+  }
+
+  /**
+   * @notice Get whether a Rewards Entity requires approval for agreements
+   * @param entity The entity address
+   * @return bool True if the entity requires approval, false otherwise
+   */
+  function getRequiresApprovalForRewardsAgreements(
+    address entity
+  ) external view returns (bool) {
+    return _requiresApproval[_entityToId[entity]];
+  }
+
+  /**
+   * @notice Set whether a Rewards Entity requires approval for agreements
+   */
+  function setRequiresApprovalForRewardsAgreements(
```

Comment:
in the design doc we specified only `requireApprovalForRewardsAgreements` with no way to undo this action, but i implemented a toggle-able function here assuming that providers may regret this decision (since they create extra work for themselves, that they may not realise)